### PR TITLE
Add initial support for beats central management

### PR DESCRIPTION
**Pull request by jalvz**
## Motivation/summary

Registers a reloadable object in Central Management so that when APM Server runs under Elastic Agent, it can receive and apply configuration from it.

In practice APM Server will only expect configuration once. This is enough for the standalone mode.

## Checklist

- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)

I have considered changes for:
- [ ] documentation
- [ ] logging (add log lines, choose appropriate log selector, etc.)
- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))
- [ ] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
- [ ] telemetry
- [ ] Elasticsearch Service (https://cloud.elastic.co)
- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)
- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)

## How to test these changes

Pull in https://github.com/elastic/beats/pull/21225 and build the Agent binary
Add an `apm` input to `elastic-agent.yml` with a well known config, eg:

```
inputs:
  - type: apm
    rum.enabled: true
    ...
``` 
Check that apm-server is running with RUM enabled

## Related issues

Required for https://github.com/elastic/apm-server/issues/4004

### CODE REVIEW

**Comment by jalvz** [comment]
Some more detailed instructions about testing:

- Check out  elastic/beats#21225
- `cd` into `x-pack/elastic-agent`
- remove the `disabled` suffix from `spec/apm-server.yml`
- `mage update` && `mage`
- `cd` into `/build/distributions/elastic-agent-7.9.0-linux-x86_64/data/elastic-agent-{HASH}`
- copy the right compiled files from this branch to `downloads` and `install`
- modify `elastic-agent.yml` as follows
```
outputs:
  default:
    type: elasticsearch
    hosts: [127.0.0.1:9200]
    username: elastic
    password: changeme
  admin:
    type: elasticsearch
    hosts: [127.0.0.1:9200]
    username: admin
    password: changeme

inputs:
  - type: apm
    rum.enabled: true    
    data_stream.namespace: prod
    use_output: admin
```
- run `elastic-agent` 
- find apm-server log at `logs/admin/apm-server.json.log`
- `POST` a valid payload to `http://localhost:8200/intake/v2/rum/events` and check the indexed document in Elasticsearch

File: beater/beater.go [hunk 1]
```diff
@@ -136,10 +114,42 @@ type beater struct {
...
+var once sync.Once
+
 // Run runs the APM Server, blocking until the beater's Stop method is called,
 // or a fatal error occurs.
 func (bt *beater) Run(b *beat.Beat) error {
 
+	done := make(chan struct{})
+
+	var reloadable = func() reload.Reloadable {
+		return reload.ReloadableFunc(func(ucfg *reload.ConfigWithMeta) error {
```
Comment (axw) [inline beater/beater.go:126]:
```suggestion
	reloadable := reload.ReloadableFunc(func(ucfg *reload.ConfigWithMeta) error {
```

Then you can just use it directly in Register, i.e. `reload.Register.MustRegister("inputs", reloadable)`

File: beater/beater.go [hunk 2]
```diff
@@ -136,10 +114,42 @@ type beater struct {
...
+			// during startup. This might change when APM Server is included in Fleet
+			once.Do(func() {
+				defer close(done)
+				var cfg *config.Config
+				cfg, err = config.NewConfig(ucfg.Config, elasticsearchOutputConfig(b))
+				if err != nil {
+					bt.logger.Warn("Could not parse configuration from Elastic Agent ", err)
+					return
+				}
+				bt.config = cfg
```
Comment (axw) [inline beater/beater.go:138]: should we also be setting `bt.rawConfig = ucfg.Config` here?

**Review by axw — COMMENTED** [review]
Looks good, just one thing.

I suppose we should start looking at shuffling things around to use `cfgfile.RunnerList`, like heartbeat?

https://github.com/elastic/beats/blob/cb624cfd18ac07e06e633a3d960d0926bd759730/heartbeat/beater/heartbeat.go#L140-L143

That would take care of starting/stopping servers in response to config changes. Not needed for now though.

File: beater/beater.go [hunk 3]
```diff
@@ -136,10 +114,40 @@ type beater struct {
 	stopped    bool
 }
 
+var once sync.Once
```
Comment (axw) [inline beater/beater.go:117]: Sorry, I missed one thing: can this be moved into the Run function, so it's close to where it's used?

```go
var reloadOnce sync.Once
var reloadable = reload.ReloadableFunc(func(ucfg *reload.ConfigWithMeta) error {
    ...
}
```

**Review by axw — APPROVED** [review]
LGTM, just one more small thing

File: beater/beater.go (hunk 3 above)
Comment (jalvz) [reply beater/beater.go:117]: ah yes, didn't know I could do that :+1:

**Comment by axw** [comment]
Seeing as https://github.com/elastic/beats/pull/21225 is not in the 7.10 branch, I'm not sure if or how we can test this with 7.10-BC1. I could cherry-pick into the elastic-agent 7.10 branch, but I think that would defeat the purpose of manual testing. I propose we push testing this back to 7.11, when the required elastic-agent bits are available.

**Comment by jalvz** [comment]
ah didn't think of that, sgtm

**Comment by axw** [comment]
Elasticsearch and Kibana setup like in https://github.com/elastic/apm-server/pull/4542#issuecomment-748872123, with Elastic Agent running outside of Docker so I could modify `agent.download.sourceURI` (couldn't figure out if/how we could do that inside a Docker container, see https://github.com/elastic/apm-integration-testing/pull/1011#issuecomment-748885938). I encountered some issues with elastic-agent fetching apm-server (opened https://github.com/elastic/beats/pull/23235), and ended up downloading the apm-server artifact and manually placing it in the downloads directory.

I edited the APM integration to set `host` to `:8200`, and left RUM as disabled, and secret token "disabled" (should be empty string, see https://github.com/elastic/apm-server/issues/4572). APM Server logs:

```
12:07:36.493 elastic_agent.apm_server [elastic_agent.apm_server][info] Listening on: [::]:8200
12:07:36.493 elastic_agent.apm_server [elastic_agent.apm_server][info] RUM endpoints disabled.
12:07:36.493 elastic_agent.apm_server [elastic_agent.apm_server][warn] Secret token is set, but SSL is not enabled.
12:07:36.493 elastic_agent.apm_server [elastic_agent.apm_server][info] SSL disabled.
```

Because of #4572 the secret token is literally set to the string "false".

```
$ curl -f http://127.0.0.2:8200
$ curl -f -H "Authorization: Bearer false" http://127.0.0.2:8200
{
  "build_date": "2020-12-17T00:20:29Z",
  "build_sha": "f3f065860d36e013c8f78a8fdb5b6b5b6207261c",
  "version": "7.11.0"
}
```

**Pull request merged**